prov/efa/src/rdm/rxr_pkt_entry.c

```diff
@@ -364,82 +364,94 @@ void rxr_pkt_entry_append(struct rxr_pkt_entry *dst,
  * @brief Populate pkt_entry->ibv_send_wr with the information stored in pkt_entry,
  * and send it out
  *
- * @param[in] ep	rxr endpoint
- * @param[in] pkt_entry	packet entry to be sent
- * @param[in] flags	flags to be applied to the send operation
+ * @param[in] ep		efa RDM endpoint
+ * @param[in] pkt_entry_vec	an array of packet entries to be sent
+ * @param[in] pkt_entry_cnt	number of packet entries to be sent
+ * @param[in] flags		flags to be applied to the send operation
  * @return		0 on success
  * 			On error, a negative value corresponding to fabric errno
  */
-ssize_t rxr_pkt_entry_send(struct efa_rdm_ep *ep, struct rxr_pkt_entry *pkt_entry,
-			   uint64_t flags)
+ssize_t rxr_pkt_entry_sendv(struct efa_rdm_ep *ep,
+			    struct rxr_pkt_entry **pkt_entry_vec,
+			    int pkt_entry_cnt,
+			    uint64_t flags)
 {
-	assert(pkt_entry->pkt_size);
 
 	struct efa_rdm_peer *peer;
-	struct rxr_pkt_sendv *send = pkt_entry->send;
+	struct rxr_pkt_sendv *send;
+	struct rxr_pkt_entry *pkt_entry;
 	struct ibv_send_wr *bad_wr, *send_wr;
 	struct ibv_sge *sge;
 	int ret, total_len;
 	struct efa_conn *conn;
+	int i;
 
-	/* EFA device supports a maximum of 2 iov/SGE
-	 */
-	assert(send->iov_count <= 2);
+	for (i = 0; i < pkt_entry_cnt; ++i) {
+		pkt_entry = pkt_entry_vec[i];
+		assert(pkt_entry->pkt_size);
 
-	peer = efa_rdm_ep_get_peer(ep, pkt_entry->addr);
-	assert(peer);
-	if (peer->flags & EFA_RDM_PEER_IN_BACKOFF)
-		return -FI_EAGAIN;
-
-	conn = efa_av_addr_to_conn(ep->base_ep.av, pkt_entry->addr);
-	assert(conn && conn->ep_addr);
-
-	assert(send);
-	if (send->iov_count == 0) {
-		send->iov_count = 1;
-		send->iov[0].iov_base = pkt_entry->wiredata;
-		send->iov[0].iov_len = pkt_entry->pkt_size;
-		send->desc[0] = pkt_entry->mr;
-	}
+		send = pkt_entry->send;
+
+		/* EFA device supports a maximum of 2 iov/SGE
+		 */
+		assert(send->iov_count <= 2);
+
+		peer = efa_rdm_ep_get_peer(ep, pkt_entry->addr);
+		assert(peer);
+		if (peer->flags & EFA_RDM_PEER_IN_BACKOFF)
+			return -FI_EAGAIN;
```

Comment:
Returning error in this loop is problematic.

It reads to me that, it is possible the function leaves `ep->base_ep.xmit_more_wr_tail` in an unknown state if it returns an error, e.g. we appended 10 send_wr to it but then returns an error to the caller.

Am I missing anything? I'm trying to understand the atomicity of this function.

Reply:
You are right this should have been moved to the before the loop.

Reply:
addressed in new revision.